### ISSUE TITLE
Haskell support

### ISSUE THREAD

**Issue by hasufell** [dependabot/dependabot-core#2745]
The Haskell ecosystem often suffers from outdated package bounds. Various attempts have been made to improve the situation, but ultimately developers need a convenient method to update their bounds.

Haskell uses [PVP](https://pvp.haskell.org/) for versioning (which is somewhat similar to semver). It then uses [Cabal](https://github.com/haskell/cabal/) for package and version management (similar to rusts cargo). Packages are shared and uploaded on [hackage](http://hackage.haskell.org/).

I think the main hurdle of adding support here is that the cabal file format is custom. Obviously there exists a haskell parser in the [Cabal library](http://hackage.haskell.org/package/Cabal). There also seems to be *some* [documentation on the file format](https://cabal.readthedocs.io/en/latest/buildinfo-fields-reference.html). I'm not sure if that's enough to re-implement a parser or if a regexp solution would be possible.

**Comment by jurre** [comment]
Thanks for opening this issue @hasufell.

Unfortunately it might be a while before our team has the capacity to pick up a new ecosystem like Haskell.

However, when we do get around to it, I think our preferred way to deal with parsing the Cabal files would be to write a [native helper, like this one](https://github.com/dependabot/dependabot-core/blob/7d57a9d256d0b6d91c94b746794b5ffacce99fad/go_modules/helpers/updater/main.go) in Haskell, and shell out to that, instead of trying to parse the file and keeping up with any changes made to the format.

**Comment by expipiplus1** [comment]
It's worth noting that many Haskell packages are described in YAML using the schema described here: https://github.com/sol/hpack

**Comment by hasufell** [comment]
> It's worth noting that many Haskell packages are described in YAML using the schema described here: https://github.com/sol/hpack

Hpack is an optional tool that *generates* cabal files. It isn't a standard.

**Comment by KiaraGrouwstra** [comment]
[related thread](https://www.reddit.com/r/haskell/comments/ev41o7/dependabot_for_haskell/)

**Comment by KiaraGrouwstra** [comment]
for transparency, I opened a PR for my wip attempt to add haskell support at #3195.

**Comment by brightly-salty** [comment]
With regard to the hpack issue, I think if it would be annoying to have the cabal updated without the package.yaml being updated, because it will be overwritten the next time you build. I feel it would be a good first step to focus on `package.yaml` files first before or in addition to `.cabal` files being added and updated.

**Comment by brightly-salty** [comment]
I agree that everyone uses cabal files, but I think that updating cabal files without updating package.yaml files would break many people's workflows. For example, I don't keep my cabal files in git because it results in needless noisey changes whenever a file is deleted or created, and instead just keep my package.yaml in git. Maybe it would be good to have a configuration option with this, but I think universally updating cabal files without regard to package.yaml files would deter me from using dependabot on my projects, and may do so for others as well.

**Comment by hasufell** [comment]
> For example, I don't keep my cabal files in git

That's bad. Please commit them. Otherwise you're breaking cabal users ability to use your repository via [source-repository-package](https://cabal.readthedocs.io/en/3.4/cabal-project.html#specifying-packages-from-remote-version-control-locations) stanza.

Note: I'm not advocating against supporting package.yaml, but I think it's an unreasonable proposal to support a 3rd party tool before the canoncial package definition format.

**Comment by Anton-Latukha** [comment]
https://github.com/dependabot/dependabot-core/issues/2745#issuecomment-845236742
>  but I think that updating cabal files without updating package.yaml files would break many people's workflows.

There is some puzzlement. There is the implicit assumption that people that use HPack would somehow be taken hostage & forced to merge Dependabot automation for Cabal in their repositories. For Dependabot Cabal support to break their HPack-configured projects.

I am an HNix developer - I do not conceive to ask to support Cabal Nix & NixPkgs integration setup before regular Cabal.

Every packaging tool in the Haskell system depends on a minimal Cabal <-> Hackage infrastructure loop.

---

<details><summary> About not including Cabal in source </summary>

</p>

Not including Cabal description into source tree, but then it somehow magically appears for Hackage upload. & now pipeline & maintainers downstream are puzzled - why tarball builds but the source from the official repository does not & why there is a file difference in those, they need to decipher what gets injected behind the scenes. It also means the project is not reproducible, so works against efforts: https://reproducible-builds.org/projects/ & https://tests.reproducible-builds.org/debian/reproducible.html.

As the whole Haskell ecosystem (& even NixPkgs `haskellPackage` set - depend on Cabal package description format. Currently.) - because everything depends on Cabal - please do not exclude Cabal descriptions from the source code, & if using HPack - just treat Cabal file as a state update details file that needs to be included in source code commits from time to time.

</details>

**Comment by Anton-Latukha** [comment]
> The Haskell ecosystem often suffers from outdated package bounds.

I resonate strongly with that idea. Ecosystem would benefit immensely just by having a default way of notifying maintainers & suggesting updates. Some community members have ~100+ packages, & there always would be such things, for old history to not repeat itself - the default way of automation would be tremendously beneficial. If someone up the stack can merge new `base` bound.

---

Because of Haskell strong typing & referential transparency - FFI with Haskell can give the highest guarantees of stability.

I noticed that Dependabot formerly had docs on FFI. & it currently seems to require https://github.com/ffi/ffi in package descriptions in the bundler. If GitHub can not provide Haskell support - the next best option for both sides is community-managed implementation through FFI.

**Comment by jurre** [comment]
Fwiw we're starting to think about a way to push down a lot of the functionality that is currently implemented in dependabot-core into the package ecosystem itself, leaving dependabot-core as a thing wrapper around the native functionality. While this is still mostly in an idea/prototyping stage, I do think that once something like that lands it would open up support for more ecosystems again.

Once we have something substantial to share I'll loop back in this thread.

**Comment by eborden** [comment]
@jurre I'm curious if any of those ideas have crystallized? Are they documented anywhere?

**Comment by jurre** [comment]
> @jurre I'm curious if any of those ideas have crystallized? Are they documented anywhere?

Not quite there yet but we're working on it

**Comment by domenkozar** [comment]
@jurre any updates?

**Comment by domenkozar** [comment]
Also worth keeping an eye on: https://github.com/renovatebot/renovate/issues/8187

**Comment by ArturGajowy** [comment]
@jurre any updates on supporting more ecosystems / outsourcing the support?

**Comment by mihaimaruseac** [comment]
Any update here? I was mistakenly thinking that Haskell support was already implemented :(

**Comment by ysangkok** [comment]
I have submitted a PR to add Cabal support to Renovate, please have a look:

* https://github.com/renovatebot/renovate/pull/33142

This currently only supports version constraints of the form `>=X.Y && <V.W`. I'd be open to suggestions for next steps, please just tag me in the Renovate discussion system.
